---
title: "Example: germline"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Example: germline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# NOTE: This article is a work in progress - not done yet.

## Introduction


This vignette will demonstrate how to use the `mdlvalr` package functions to generate a concordance analysis of data generated by the "aws_v6.0_gatk38germline" pipeline.



Load required packages.

```{r, warning = FALSE, message = FALSE}
library(mdlvalr)
library(tidyverse)
library(glue)
library(openxlsx)
library(plyranges)
```


## Get input data. Use the provided example data that comes with the `mdlvalr` R package. 

Load example data objects


```{r}
# Assign package lazy data object to a env variable
cds_table <- refseq_gff_unnest_with_chrom_cds
```

Get the path to package extra data dir.

```{r}
extdata_dir <- system.file("extdata", package = "mdlvalr", mustWork = TRUE)
```

Find the real file paths of variant and coverage files.


```{r}
var_Dec01 <- Sys.glob(glue("{extdata_dir}/aws_v6.0_gatk38germline/*_01Dec2010_*.vcf"))
var_Dec02 <- Sys.glob(glue("{extdata_dir}/aws_v6.0_gatk38germline/*_02Dec2010_*.vcf"))
cov_Dec01 <- Sys.glob(glue("{extdata_dir}/aws_v6.0_gatk38germline/*_01Dec2010_*.tab"))
cov_Dec02 <- Sys.glob(glue("{extdata_dir}/aws_v6.0_gatk38germline/*_02Dec2010_*.tab"))
```


Create a sample sheet of comparisons


```{r}
sample_sheet <- tibble::tibble(
    var_path_1 = var_Dec01,
    var_path_2 = var_Dec02,
    cov_path_1 = cov_Dec01,
    cov_path_2 = cov_Dec02 
    ) %>%
    dplyr::mutate(comparison_name_1 = str_remove(basename(var_path_1), "_MSOC191_MDL55455.*$")) %>%
    dplyr::mutate(comparison_name_2 = str_remove(basename(var_path_2), "_MSOC191_MDL55455.*$")) %>%
    dplyr::mutate(comparison_name = glue("{comparison_name_1}_vs_{comparison_name_2}"))
```



Get a list of known genes, which might be helpful for filtering later


```{r}
gene_list <- read_tsv(glue("{extdata_dir}/aws_v6.0_gatk38germline/exons_MSOC.bed.geneList.tab"),
    col_names = "gene",
    col_types = cols(.default = "c"))
```




## Run concordance

```{r}
mdlvalr_list <- get_data(sample_sheet, pipeline = "germline")


# NEEDS DEVELOPMENT
# E.g. keep only vars that PASS and percent coverage of exons covered at 20x >= 1
# mdlvalr_list <- add_flags(
#     mdlvalr_list,
#     cds_table = cds_table,
#     var_pass_fail_logic = "FILTER == 'PASS'",
#     cov_pass_fail_logic = "exons_20x >= 1"
# )
#
# mdlvalr_list <- compare_vars(mdlvalr_list)
# mdlvalr_list <- add_var_labels(mdlvalr_list)
# mdlvalr_list <- filter_data(mdlvalr_list)
# mdlvalr_list <- get_stats(mdlvalr_list)
```




## Create output files


```{r}
# NEEDS DEVELOPMENT
# export_excel(mdlvalr_list, filename_prefix = "")
```





## Explore the `mdlvalr_list` object

View the list object

```{r}
# Show the list-of-lists object
glimpse(mdlvalr_list)
```



Print the summary stats table

```{r}
# NEEDS DEVELOPMENT
# mdlvalr_list$summary %>%
#     print(width = Inf, n = 6)
```

Names of the filtered table outputs for the first comparison

```{r}
# NEEDS DEVELOPMENT
# names(mdlvalr_list$comparisons[[1]]$filtered_data)
```


