<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mdlvalr">
<title>About the package • mdlvalr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="About the package">
<meta property="og:description" content="mdlvalr">
<meta property="og:image" content="https://pages.github.umn.edu/msi/mdlvalr/index.html/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mdlvalr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/about.html">About the package</a>
    <a class="dropdown-item" href="../articles/germline.html">Example: germline</a>
    <a class="dropdown-item" href="../articles/hybcap.html">Example: hybcap</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.umn.edu/msi/mdlvalr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>About the package</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.umn.edu/msi/mdlvalr/blob/HEAD/vignettes/about.Rmd" class="external-link"><code>vignettes/about.Rmd</code></a></small>
      <div class="d-none name"><code>about.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pages.github.umn.edu/msi/mdlvalr" class="external-link">mdlvalr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The MDL has multiple variant calling assays that are under repeated wet-lab and bioinformatics pipeline development. We need a good way to compare variant calling results after changes are made to the process. Therefore, we have developed a “concordance” framework that could be applied to the majority of our pipelines.</p>
</div>
<div class="section level2">
<h2 id="the-problem">The problem<a class="anchor" aria-label="anchor" href="#the-problem"></a>
</h2>
<p>Each of the various pipelines have slightly different output files (some return VCFs, some return Excel tables, etc.). Therefore, we would like to create a single unified concordance method to compare A vs. B, for all pipelines – but this is not really possible.</p>
</div>
<div class="section level2">
<h2 id="a-potential-solution">A potential solution<a class="anchor" aria-label="anchor" href="#a-potential-solution"></a>
</h2>
<p>In one way, we could build multiple different concordance scripts that each handle specific pipelines. Alternatively, we could build a basic framework of concordance analysis that can be adjusted for each of the pipelines, but try to standardize each of the main steps.</p>
<p>For example, the basic approach of <code>mdlvalr</code> is to exploit an R programming language feature, called function “Methods”. We can build one function (e.g. <code><a href="../reference/compare_vars.html">compare_vars()</a></code>) that will “compare variant tables”, but create different methods that correspond to the pipeline of interest. This idea is exploited widely in R. For example, when you execute <code>print(object)</code> in R, it looks to see what <code>class</code> the object is, and returns a class-specific print out. This is how R knows how to print a linear model fit vs a data frame (<code>print(fit)</code> vs <code>print(df)</code>) to the screen using the same <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> function. Thus, <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> is a generic function in R that employs different methods. To list all the methods available to the generic print function, you can run: <code>methods(print)</code>. And if you load certain libraries (e.g. <code>tidyverse</code>), many more methods will be avaialble.</p>
<p>For <code>mdlvalr</code> we can design the basic concordance steps as generic functions, but write pipeline specific methods, when needed. For example, in R code, the generic function name would be <code>step1</code> and the pipeline specific method name would be <code>step1.pipeline1</code>. Using this approach, if <code>class(object)</code> includes the class name “pipeline1”, then we can run <code>step1(object)</code> and the step1 function will automatically apply the pipeline1 method.</p>
<p>Alternatively, we could simply build pipeline specific functions (e.g. <code>step1_pipeline1()</code>) for each and every step and pipeline combination. Exploiting the R generics/methods framework simply makes the code (and future code maintenance) more uniform.</p>
</div>
<div class="section level2">
<h2 id="a-general-concordance-framework">A general concordance framework<a class="anchor" aria-label="anchor" href="#a-general-concordance-framework"></a>
</h2>
<p>For each step below (except <code><a href="../reference/get_data.html">get_data()</a></code>), a generic function was created. The generic function will point to a pipeline-specific “method” to be used (based on the class of the object).</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/get_data.html">get_data()</a></code>
<ul>
<li>This is not an R generic function</li>
<li>The purpose of this function is to take a sample sheet of comparisons as input and capture (read-in) all the necessary data to complete the analysis.</li>
<li>The <code>pipeline</code> needs to be specified for this function.</li>
<li>A standard R list object is output from this function, but the list also has a new R class added to its attributes: the name of the pipeline (e.g. <code>hybcap</code>).</li>
</ul>
</li>
<li>
<code><a href="../reference/add_flags.html">add_flags()</a></code>
<ul>
<li>The purpose of this function is to add new columns to the variant and coverage tables.</li>
<li>A coding sequence (CDS) table is used to determine whether a variant in the sample table falls within an exon or not. This results in a new column (<code>within_refseq_cds</code>) being added to the table with “yes” or “no” values.</li>
<li>Frequently, the clinicians only mostly care about concordance of vars or coverage regions based on some thresholds. These thresholds can be ad hoc. Therefore, we can specify pipeline-specific R filtering logic to the <code>var_pass_fail_logic</code> or <code>cov_pass_fail_logic</code> parameters. If the variant or coverage row passes the logic test (i.e. <code>TRUE</code>), then a new column is added to the table indicating “pass” or “fail”.</li>
<li>No variants or coverage rows are removed from any tables.</li>
</ul>
</li>
<li>
<code><a href="../reference/compare_vars.html">compare_vars()</a></code>
<ul>
<li>The purpose of this function is to compare variant or coverage tables from each comparison group.</li>
<li>The function runs <code>dylyr::full_join()</code> (among other joins) by a set of common columns (e.g. CHROM, REF, ALT, etc.). All possible splits of the data are captured into sub-tables and saved to the R list object.</li>
</ul>
</li>
<li>
<code><a href="../reference/add_var_labels.html">add_var_labels()</a></code>
<ul>
<li>The purpose of this function is to add new columns to the “compared” data tables.</li>
<li>Using the pass/fail labels (from <code><a href="../reference/add_flags.html">add_flags()</a></code>) and along with their group membership (e.g. <code>vars_in_s1_not_in_s2</code>), we can further label variants as passing or failing in one or both samples being compared.</li>
<li>No variants are removed from any tables</li>
</ul>
</li>
<li>
<code><a href="../reference/filter_data.html">filter_data()</a></code>
<ul>
<li>Finally, after comparing tables and labeling vars/coverage rows, we can use these labels to actually filter the tables.</li>
<li>Tables can be filtered down into smaller “pass-only”, or “pass in sample1, fail in sample2” type tables.</li>
<li>These filtered tables are added to the input list object.</li>
</ul>
</li>
<li>
<code><a href="../reference/get_stats.html">get_stats()</a></code>
<ul>
<li>This function simply reports the size of the all the various tables in the list object.</li>
<li>For example: the total number of rows in the var_1 table, or the number of rows in the <code>vars_in_common_s1pass_s2pass</code> table.</li>
</ul>
</li>
<li>
<code><a href="../reference/export_excel.html">export_excel()</a></code>
<ul>
<li>For each comparison, this function writes the various tables generated above into an Excel workbook, using tabs to represent different splits of the data.</li>
<li>Alternate exports of data could be imagined as well (i.e. not excel files – maybe Rmarkdown, etc.).</li>
</ul>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="note-about-r-generics">Note about R generics<a class="anchor" aria-label="anchor" href="#note-about-r-generics"></a>
</h2>
<p>Accessing the docs for a function in R is as easy, using <code>?function_name</code>. This will work great for <code>?get_data()</code> because it is a regular function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pages.github.umn.edu/msi/mdlvalr" class="external-link">mdlvalr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="op">?</span><span class="fu"><a href="../reference/get_data.html">get_data</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">get_data</span></span>
<span><span class="co">#&gt; function (sample_sheet, pipeline) </span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;     if (!pipeline %in% c("hybcap", "germline")) {</span></span>
<span><span class="co">#&gt;         stop("ERROR: pipeline arg supplied to `get_data()` must be only one of: c('hybcap', 'germline')")</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     comparisons &lt;- list()</span></span>
<span><span class="co">#&gt;     if (pipeline == "hybcap") {</span></span>
<span><span class="co">#&gt;         for (i in seq_len(nrow(sample_sheet))) {</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$var_path_1[i])) {</span></span>
<span><span class="co">#&gt;                 var_1 &lt;- openxlsx::read.xlsx(xlsxFile = sample_sheet$var_path_1[i]) %&gt;% </span></span>
<span><span class="co">#&gt;                   as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find var_path_1 file: {sample_sheet$var_path_1[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$var_path_2[i])) {</span></span>
<span><span class="co">#&gt;                 var_2 &lt;- openxlsx::read.xlsx(xlsxFile = sample_sheet$var_path_2[i]) %&gt;% </span></span>
<span><span class="co">#&gt;                   as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find var_path_2 file: {sample_sheet$var_path_2[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$cov_path_1[i])) {</span></span>
<span><span class="co">#&gt;                 cov_1 &lt;- openxlsx::read.xlsx(xlsxFile = sample_sheet$cov_path_1[i], </span></span>
<span><span class="co">#&gt;                   sheet = 2) %&gt;% as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find cov_path_1 file: {sample_sheet$cov_path_1[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$cov_path_2[i])) {</span></span>
<span><span class="co">#&gt;                 cov_2 &lt;- openxlsx::read.xlsx(xlsxFile = sample_sheet$cov_path_2[i], </span></span>
<span><span class="co">#&gt;                   sheet = 2) %&gt;% as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find cov_path_2 file: {sample_sheet$cov_path_2[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             comparisons[[i]] &lt;- list(input_data = list(var_1 = var_1, </span></span>
<span><span class="co">#&gt;                 var_2 = var_2, cov_1 = cov_1, cov_2 = cov_2))</span></span>
<span><span class="co">#&gt;         }</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     else if (pipeline == "germline") {</span></span>
<span><span class="co">#&gt;         for (i in seq_len(nrow(sample_sheet))) {</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$var_path_1[i])) {</span></span>
<span><span class="co">#&gt;                 var_1 &lt;- readr::read_tsv(sample_sheet$var_path_1[i], </span></span>
<span><span class="co">#&gt;                   comment = "#", col_types = cols(.default = "c"), </span></span>
<span><span class="co">#&gt;                   col_names = c("CHROM", "POS", "ID", "REF", </span></span>
<span><span class="co">#&gt;                     "ALT", "QUAL", "FILTER", "INFO", "FORMAT", </span></span>
<span><span class="co">#&gt;                     "SAMPLE")) %&gt;% as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find var_path_1 file: {sample_sheet$var_path_1[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$var_path_2[i])) {</span></span>
<span><span class="co">#&gt;                 var_2 &lt;- readr::read_tsv(sample_sheet$var_path_2[i], </span></span>
<span><span class="co">#&gt;                   comment = "#", col_types = cols(.default = "c"), </span></span>
<span><span class="co">#&gt;                   col_names = c("CHROM", "POS", "ID", "REF", </span></span>
<span><span class="co">#&gt;                     "ALT", "QUAL", "FILTER", "INFO", "FORMAT", </span></span>
<span><span class="co">#&gt;                     "SAMPLE")) %&gt;% as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find var_path_2 file: {sample_sheet$var_path_2[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$cov_path_1[i])) {</span></span>
<span><span class="co">#&gt;                 cov_1 &lt;- readr::read_tsv(sample_sheet$cov_path_1[i], </span></span>
<span><span class="co">#&gt;                   comment = "#", col_types = cols(.default = "c"), </span></span>
<span><span class="co">#&gt;                   col_names = c("chr", "start", "end", "bases_15x", </span></span>
<span><span class="co">#&gt;                     "bases_20x", "exons_15x", "exons_20x", "exon")) %&gt;% </span></span>
<span><span class="co">#&gt;                   as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find cov_path_1 file: {sample_sheet$cov_path_1[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             if (file.exists(sample_sheet$cov_path_2[i])) {</span></span>
<span><span class="co">#&gt;                 cov_2 &lt;- readr::read_tsv(sample_sheet$cov_path_2[i], </span></span>
<span><span class="co">#&gt;                   comment = "#", col_types = cols(.default = "c"), </span></span>
<span><span class="co">#&gt;                   col_names = c("chr", "start", "end", "bases_15x", </span></span>
<span><span class="co">#&gt;                     "bases_20x", "exons_15x", "exons_20x", "exon")) %&gt;% </span></span>
<span><span class="co">#&gt;                   as_tibble()</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             else {</span></span>
<span><span class="co">#&gt;                 stop(glue("ERROR: cannot find cov_path_2 file: {sample_sheet$cov_path_2[i]}"))</span></span>
<span><span class="co">#&gt;             }</span></span>
<span><span class="co">#&gt;             comparisons[[i]] &lt;- list(input_data = list(var_1 = var_1, </span></span>
<span><span class="co">#&gt;                 var_2 = var_2, cov_1 = cov_1, cov_2 = cov_2))</span></span>
<span><span class="co">#&gt;         }</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     else {</span></span>
<span><span class="co">#&gt;         stop(glue("ERROR: There is no method for get_data for pipeline: {pipeline}"))</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     names(comparisons) &lt;- sample_sheet$comparison_name</span></span>
<span><span class="co">#&gt;     mdlvalr_list &lt;- list(pipeline = pipeline, sample_sheet = sample_sheet, </span></span>
<span><span class="co">#&gt;         summary = NULL, comparisons = comparisons)</span></span>
<span><span class="co">#&gt;     class(mdlvalr_list) &lt;- c(pipeline, class(mdlvalr_list))</span></span>
<span><span class="co">#&gt;     return(mdlvalr_list)</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; &lt;bytecode: 0x55154e8&gt;</span></span>
<span><span class="co">#&gt; &lt;environment: namespace:mdlvalr&gt;</span></span></code></pre></div>
<p>However, doing this with generic functions leaves you wondering more! Instead, for generics, try the following to find the code behind R generic functions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="fu"><a href="../reference/compare_vars.html">compare_vars</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">compare_vars</span></span>
<span><span class="co">#&gt; function (mdlvalr_list, pipeline = NULL, match_by_colnames = NULL, </span></span>
<span><span class="co">#&gt;     suffix_names = c(".sample1", ".sample2")) </span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;     UseMethod("compare_vars")</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; &lt;bytecode: 0x7be0298&gt;</span></span>
<span><span class="co">#&gt; &lt;environment: namespace:mdlvalr&gt;</span></span>
<span></span>
<span><span class="co"># List the available methods for this generic function</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/methods.html" class="external-link">methods</a></span><span class="op">(</span><span class="va">compare_vars</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] compare_vars.hybcap*</span></span>
<span><span class="co">#&gt; see '?methods' for accessing help and source code</span></span>
<span></span>
<span><span class="co"># Show the source code for the method: "hybcap"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/getAnywhere.html" class="external-link">getAnywhere</a></span><span class="op">(</span><span class="va">compare_vars.hybcap</span><span class="op">)</span></span>
<span><span class="co">#&gt; A single object matching 'compare_vars.hybcap' was found</span></span>
<span><span class="co">#&gt; It was found in the following places</span></span>
<span><span class="co">#&gt;   registered S3 method for compare_vars from namespace mdlvalr</span></span>
<span><span class="co">#&gt;   namespace:mdlvalr</span></span>
<span><span class="co">#&gt; with value</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; function (mdlvalr_list, pipeline = "hybcap", match_by_colnames = c("CHROM", </span></span>
<span><span class="co">#&gt;     "POS", "REF", "ALT", "GENE", "EXON"), suffix_names = c(".sample1", </span></span>
<span><span class="co">#&gt;     ".sample2")) </span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;     for (i in seq_along(names(mdlvalr_list$comparisons))) {</span></span>
<span><span class="co">#&gt;         var_1 &lt;- mdlvalr_list$comparisons[[i]]$flagged_data$var_1</span></span>
<span><span class="co">#&gt;         var_2 &lt;- mdlvalr_list$comparisons[[i]]$flagged_data$var_2</span></span>
<span><span class="co">#&gt;         full_join &lt;- dplyr::full_join(var_1, var_2, by = match_by_colnames, </span></span>
<span><span class="co">#&gt;             suffix = suffix_names)</span></span>
<span><span class="co">#&gt;         vars_in_common &lt;- dplyr::inner_join(var_1, var_2, by = match_by_colnames, </span></span>
<span><span class="co">#&gt;             suffix = suffix_names)</span></span>
<span><span class="co">#&gt;         vars_in_common_from_s1 &lt;- dplyr::semi_join(var_1, var_2, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_common_from_s2 &lt;- dplyr::semi_join(var_2, var_1, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_s1_not_in_s2 &lt;- dplyr::anti_join(var_1, var_2, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_s2_not_in_s1 &lt;- dplyr::anti_join(var_2, var_1, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         mdlvalr_list$comparisons[[i]]$compared_data &lt;- list(full_join = full_join, </span></span>
<span><span class="co">#&gt;             vars_in_common = vars_in_common, vars_in_common_from_s1 = vars_in_common_from_s1, </span></span>
<span><span class="co">#&gt;             vars_in_common_from_s2 = vars_in_common_from_s2, </span></span>
<span><span class="co">#&gt;             vars_in_s1_not_in_s2 = vars_in_s1_not_in_s2, vars_in_s2_not_in_s1 = vars_in_s2_not_in_s1)</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     class(mdlvalr_list) &lt;- class(mdlvalr_list)</span></span>
<span><span class="co">#&gt;     return(mdlvalr_list)</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; &lt;bytecode: 0x5ffeeb8&gt;</span></span>
<span><span class="co">#&gt; &lt;environment: namespace:mdlvalr&gt;</span></span>
<span><span class="fu">mdlvalr</span><span class="fu">:::</span><span class="va"><a href="../reference/compare_vars.html">compare_vars.hybcap</a></span></span>
<span><span class="co">#&gt; function (mdlvalr_list, pipeline = "hybcap", match_by_colnames = c("CHROM", </span></span>
<span><span class="co">#&gt;     "POS", "REF", "ALT", "GENE", "EXON"), suffix_names = c(".sample1", </span></span>
<span><span class="co">#&gt;     ".sample2")) </span></span>
<span><span class="co">#&gt; {</span></span>
<span><span class="co">#&gt;     for (i in seq_along(names(mdlvalr_list$comparisons))) {</span></span>
<span><span class="co">#&gt;         var_1 &lt;- mdlvalr_list$comparisons[[i]]$flagged_data$var_1</span></span>
<span><span class="co">#&gt;         var_2 &lt;- mdlvalr_list$comparisons[[i]]$flagged_data$var_2</span></span>
<span><span class="co">#&gt;         full_join &lt;- dplyr::full_join(var_1, var_2, by = match_by_colnames, </span></span>
<span><span class="co">#&gt;             suffix = suffix_names)</span></span>
<span><span class="co">#&gt;         vars_in_common &lt;- dplyr::inner_join(var_1, var_2, by = match_by_colnames, </span></span>
<span><span class="co">#&gt;             suffix = suffix_names)</span></span>
<span><span class="co">#&gt;         vars_in_common_from_s1 &lt;- dplyr::semi_join(var_1, var_2, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_common_from_s2 &lt;- dplyr::semi_join(var_2, var_1, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_s1_not_in_s2 &lt;- dplyr::anti_join(var_1, var_2, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         vars_in_s2_not_in_s1 &lt;- dplyr::anti_join(var_2, var_1, </span></span>
<span><span class="co">#&gt;             by = match_by_colnames)</span></span>
<span><span class="co">#&gt;         mdlvalr_list$comparisons[[i]]$compared_data &lt;- list(full_join = full_join, </span></span>
<span><span class="co">#&gt;             vars_in_common = vars_in_common, vars_in_common_from_s1 = vars_in_common_from_s1, </span></span>
<span><span class="co">#&gt;             vars_in_common_from_s2 = vars_in_common_from_s2, </span></span>
<span><span class="co">#&gt;             vars_in_s1_not_in_s2 = vars_in_s1_not_in_s2, vars_in_s2_not_in_s1 = vars_in_s2_not_in_s1)</span></span>
<span><span class="co">#&gt;     }</span></span>
<span><span class="co">#&gt;     class(mdlvalr_list) &lt;- class(mdlvalr_list)</span></span>
<span><span class="co">#&gt;     return(mdlvalr_list)</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; &lt;bytecode: 0x5ffeeb8&gt;</span></span>
<span><span class="co">#&gt; &lt;environment: namespace:mdlvalr&gt;</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Todd Knutson, Christine O’Connor.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
