#' Compare samples using join functions
#'
#' @export
#' @param mdlvalr_list A standard `mdlvalr_list` object (as generated by `get_flags()`).
#' The class of this object can be used to automatically determine which
#' function methods will be used.
#' @param pipeline Standard name of the pipeline from which the data were derived.
#' @param match_by_colnames A vector of column names that should be present in each sample variant table. These columns will be used for joining the tables (and thus, rows with exact matching data is merged).
#' @param suffix_names A character vector of length two, that provides the suffix string to append to column names of each sample, making the colnames unique. 
#'
#' @section Methods:
#' This function is a **generic**, which means that different
#' implementations (methods) can be executed for different classes. The class
#' of the first argument will determine which method is used. See the documentation of
#' individual methods for extra arguments and differences in behaviour.
#'
#' @return Returns a standard `mdlvalr_list` object (i.e. a regular R list of lists) that contains tables as named list elements. 
#' The class of the retuned object is updated to include the name of the pipeline. This facilitates use of alternative methods in 
#' downstream functions. 
#'
#' @examples
#' class(mdlvalr_list)
#' # [1] "hybcap" "list"
#' \notrun{
#' mdlvalr_list <- compare_vars(
#'     mdlvalr_list,
#'     pipeline = "hybcap",
#'     match_by_colnames = c("CHROM", "POS", "REF", "ALT", "GENE", "EXON"),
#'     suffix_names = c(".sample1", ".sample2"))
#'}
compare_vars <- function(
    mdlvalr_list,
    pipeline = NULL,
    match_by_colnames = NULL,
    suffix_names = c(".sample1", ".sample2")) {
    UseMethod("compare_vars")
}




#' @export
#' @rdname compare_vars
compare_vars.hybcap <- function(
    mdlvalr_list,
    pipeline = "hybcap",
    match_by_colnames = c("CHROM", "POS", "REF", "ALT", "GENE", "EXON"),
    suffix_names = c(".sample1", ".sample2")) {

    for (i in seq_along(names(mdlvalr_list$comparisons))) {
        var_1 <- mdlvalr_list$comparisons[[i]]$flagged_data$var_1
        var_2 <- mdlvalr_list$comparisons[[i]]$flagged_data$var_2

        # Compare tables
        full_join <- dplyr::full_join(var_1, var_2, by = match_by_colnames, suffix = suffix_names)
        vars_in_common <- dplyr::inner_join(var_1, var_2, by = match_by_colnames, suffix = suffix_names)
        vars_in_common_from_s1 <- dplyr::semi_join(var_1, var_2, by = match_by_colnames)
        vars_in_common_from_s2 <- dplyr::semi_join(var_2, var_1, by = match_by_colnames)
        vars_in_s1_not_in_s2 <- dplyr::anti_join(var_1, var_2, by = match_by_colnames)
        vars_in_s2_not_in_s1 <- dplyr::anti_join(var_2, var_1, by = match_by_colnames)

        mdlvalr_list$comparisons[[i]]$compared_data <- list(
            full_join = full_join,
            vars_in_common = vars_in_common,
            vars_in_common_from_s1 = vars_in_common_from_s1,
            vars_in_common_from_s2 = vars_in_common_from_s2,
            vars_in_s1_not_in_s2 = vars_in_s1_not_in_s2,
            vars_in_s2_not_in_s1 = vars_in_s2_not_in_s1
        )
    }
    class(mdlvalr_list) <- class(mdlvalr_list)
    return(mdlvalr_list)
}

